### PR TITLE
Feature/nomis integration

### PR DESCRIPTION
**Pull request by bnsports**

### CODE REVIEW

File: fearless/ApplicationLayer/Services/ScamService/ScamInfoFetcher.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        let scamInfo: ScamInfo? = score.flatMap {
+            guard ($0.data?.score).or(.zero) < 0.25 else {
+                return nil
+            }
+
+            return ScamInfo(
+                name: "Nomis multi-chain score",
+                address: address,
+                type: .lowScore,
+                subtype: "Proceed with caution"
```
Comment (bnsports) [inline fearless/ApplicationLayer/Services/ScamService/ScamInfoFetcher.swift]: localization

File: fearless/ApplicationLayer/Services/SoraFiatService/SubqueryFiatInfoOperation.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import RobinHood
+import sorawallet
+import Foundation
+
+public final class SubqueryFiatInfoOperation<ResultType>: BaseOperation<ResultType> {
```
Comment (bnsports) [inline fearless/ApplicationLayer/Services/SoraFiatService/SubqueryFiatInfoOperation.swift]: possible unused

File: fearless/ApplicationLayer/Services/SoraFiatService/SoraFiatService.swift [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import Foundation
+import RobinHood
+import sorawallet
+
+protocol FiatServiceObserverProtocol: AnyObject {
```
Comment (bnsports) [inline fearless/ApplicationLayer/Services/SoraFiatService/SoraFiatService.swift]: possible unused

File: fearless/Common/DataProvider/ExistentialDeposit.swift [hunk 4]
```diff
@@ -41,46 +41,23 @@ final class ExistentialDepositService: RuntimeConstantFetching, ExistentialDepos
...
-        if
-            let existentialDeposit = chainAsset.asset.existentialDeposit,
-            let result = BigUInt(string: existentialDeposit) {
-            completion(.success(result))
-            return
-        }
-
         switch chainAsset.chainAssetType {
-        case .normal, .ormlChain, .soraAsset:
+        case .equilibrium:
```
Comment (bnsports) [inline fearless/Common/DataProvider/ExistentialDeposit.swift]: remove equilibrium case

File: fearless/Common/DataProvider/Sources/PriceDataSource.swift [hunk 5]
```diff
@@ -27,78 +30,160 @@ final class PriceDataSource: SingleValueProviderSourceProtocol {
...
             .uniq(predicate: { $0 })
+        guard priceIds.isNotEmpty else {
+            return BaseOperation.createWithResult([])
+        }
         let operation = coingeckoOperationFactory.fetchPriceOperation(for: priceIds, currencies: currencies)
         return operation
     }
 
     private func createChainlinkOperations() -> [BaseOperation<PriceData>] {
+        return []
```
Comment (bnsports) [inline fearless/Common/DataProvider/Sources/PriceDataSource.swift]: remove this. chainlink prices were disabled on config

File: fearless/Common/Extension/Foundation/NumberFormatter.swift [hunk 6]
```diff
@@ -123,4 +123,12 @@ extension NumberFormatter {
         formatter.usesGroupingSeparator = usesIntGrouping
         return formatter
     }
+
+    static var nomisHours: NumberFormatter {
```
Comment (bnsports) [inline fearless/Common/Extension/Foundation/NumberFormatter.swift]: unused

File: fearless/Common/Model/AssetVisibility.swift [hunk 7]
```diff
@@ -1,11 +1,11 @@
...
 
-struct AssetVisibility: Codable, Equatable, Hashable, Identifiable {
-    var identifier: String {
-        assetId
-    }
-
-    let assetId: String
-    var hidden: Bool
-}
+// struct AssetVisibility: Codable, Equatable, Hashable, Identifiable {
```
Comment (bnsports) [inline fearless/Common/Model/AssetVisibility.swift]: commented code

File: fearless/Common/Model/ChainRegistry/ChainAccountModel.swift [hunk 8]
```diff
@@ -1,12 +1,13 @@
...
+import SSFModels
 
-struct ChainAccountModel: Equatable, Hashable, Codable {
-    let chainId: String
-    let accountId: AccountId
-    let publicKey: Data
-    let cryptoType: UInt8
-    let ethereumBased: Bool
-}
+// struct ChainAccountModel: Equatable, Hashable, Codable {
```
Comment (bnsports) [inline fearless/Common/Model/ChainRegistry/ChainAccountModel.swift]: commented code

File: fearless/Common/Model/ChainRegistry/MetaAccountModel.swift [hunk 9]
```diff
@@ -21,6 +21,10 @@ struct MetaAccountModel: Equatable, Codable {
     let assetsVisibility: [AssetVisibility]
     let hasBackup: Bool
     let favouriteChainIds: [ChainModel.Id]
+
+    var utilsModel: SSFModels.MetaAccountModel {
+        SSFModels.MetaAccountModel(metaId: metaId, name: name, substrateAccountId: substrateAccountId, substrateCryptoType: substrateCryptoType, substratePublicKey: substratePublicKey, ethereumAddress: ethereumAddress, ethereumPublicKey: ethereumPublicKey, chainAccounts: chainAccounts, assetKeysOrder: assetKeysOrder, assetFilterOptions: [], canExportEthereumMnemonic: canExportEthereumMnemonic, unusedChainIds: unusedChainIds, selectedCurrency: selectedCurrency, networkManagmentFilter: networkManagmentFilter, assetsVisibility: assetsVisibility, zeroBalanceAssetsHidden: false, hasBackup: hasBackup, favouriteChainIds: favouriteChainIds)
```
Comment (bnsports) [inline fearless/Common/Model/ChainRegistry/MetaAccountModel.swift]: line wrapping

File: fearless/Common/Model/Currency.swift [hunk 10]
```diff
@@ -1,29 +1,29 @@
...
-        Currency(
-            id: "eur",
-            symbol: "€",
-            name: "Euro",
-            icon: "https://raw.githubusercontent.com/soramitsu/fearless-utils/android/2.0.2/icons/fiat/eur.svg",
-            isSelected: false
-        )
-    }
-}
+// struct Currency: Codable, Equatable, Hashable {
```
Comment (bnsports) [inline fearless/Common/Model/Currency.swift]: commented code

File: fearless/Common/Services/ChainRegistry/ChainRegistry.swift [hunk 11]
```diff
@@ -17,6 +17,7 @@ protocol ChainRegistryProtocol: AnyObject {
     func getConnection(for chainId: ChainModel.Id) -> ChainConnection?
     func getRuntimeProvider(for chainId: ChainModel.Id) -> RuntimeProviderProtocol?
     func getChain(for chainId: ChainModel.Id) -> ChainModel?
+    func getChainUnsafe(for chainId: ChainModel.Id) -> ChainModel?
```
Comment (bnsports) [inline fearless/Common/Services/ChainRegistry/ChainRegistry.swift]: remove this

File: fearless/Common/Services/ChainRegistry/ChainRegistry.swift [hunk 12]
```diff
@@ -226,16 +227,22 @@ final class ChainRegistry {
...
     }
 
     private func handleUpdatedEthereumChain(updatedChain: ChainModel) throws {
         guard let ethereumConnectionPool = self.ethereumConnectionPool else {
             return
         }
-        _ = try ethereumConnectionPool.setupConnection(for: updatedChain)
-        chains = chains.filter { $0.chainId != updatedChain.chainId }
-        chains.append(updatedChain)
+        DispatchQueue.global(qos: .background).async { [weak self] in
```
Comment (bnsports) [inline fearless/Common/Services/ChainRegistry/ChainRegistry.swift]: remove thread management

File: fearless/Common/Services/ChainRegistry/ChainRegistry.swift [hunk 13]
```diff
@@ -306,26 +313,30 @@ extension ChainRegistry: ChainRegistryProtocol {
...
-        readLock.concurrentlyRead {
-            guard
-                let ethereumConnectionPool = self.ethereumConnectionPool,
-                let chain = chains.first(where: { $0.chainId == chainId })
-            else {
-                return nil
-            }
-
-            return try? ethereumConnectionPool.setupConnection(for: chain)
+//        readLock.concurrentlyRead {
```
Comment (bnsports) [inline fearless/Common/Services/ChainRegistry/ChainRegistry.swift]: commented code

File: fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeProvider.swift [hunk 14]
```diff
@@ -202,6 +202,25 @@ final class RuntimeProvider {
     }
 }
 
+// <<<<<<< HEAD
```
Comment (bnsports) [inline fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeProvider.swift]: commented code

File: fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeSnapshot.swift [hunk 15]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import SSFUtils
+import SSFModels
+
+// struct RuntimeSnapshot {
```
Comment (bnsports) [inline fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeSnapshot.swift]: commented code

File: fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeSpecVersion.swift [hunk 16]
```diff
@@ -0,0 +1,45 @@
+import Foundation
+import SSFModels
+
+// enum RuntimeSpecVersion: UInt32, CaseIterable {
```
Comment (bnsports) [inline fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeSpecVersion.swift]: commented code

File: fearless/Common/Services/RuntimeRegistryService/RuntimeCoderFactory.swift [hunk 17]
```diff
@@ -0,0 +1,38 @@
+import Foundation
+import SSFUtils
+
+// protocol RuntimeCoderFactoryProtocol {
```
Comment (bnsports) [inline fearless/Common/Services/RuntimeRegistryService/RuntimeCoderFactory.swift]: commented code

File: fearless/Common/Services/RuntimeRegistryService/RuntimeRegistryServiceProtocol.swift [hunk 18]
```diff
@@ -0,0 +1,26 @@
+import Foundation
+import RobinHood
+import SSFUtils
+
+// typealias RuntimeMetadataClosure = () throws -> RuntimeMetadata
```
Comment (bnsports) [inline fearless/Common/Services/RuntimeRegistryService/RuntimeRegistryServiceProtocol.swift]: commented code

File: fearless/Common/View/AccountScore/AccountScoreView.swift [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+        skeletonView?.stopSkrulling()
+        skeletonView?.removeFromSuperview()
+        skeletonView = nil
+
+        starView.alpha = 1.0
+    }
+
+    private func setupSkeleton() {
+        let spaceSize = CGSize(width: 32, height: 15)
```
Comment (bnsports) [inline fearless/Common/View/AccountScore/AccountScoreView.swift]: size to constants

File: fearless/Common/View/AccountScore/AccountScoreView.swift [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+            starView.rating = 5
+        case .medium:
+            starView.rating = 2.5
+        case .low:
+            starView.rating = 0
+        }
+    }
+
+    func bindEmptyViewModel() {
+        starView.text = "N/A"
```
Comment (bnsports) [inline fearless/Common/View/AccountScore/AccountScoreView.swift:78]: localization ?

File: fearless/Common/ViewModel/AccountScore/AccountScoreViewModel.swift [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+import SoraKeystore
+import SSFModels
+
+enum AccountScoreRate {
+    case low
+    case medium
+    case high
+
+    init(from score: Decimal) {
+        if score < 0.25 {
```
Comment (bnsports) [inline fearless/Common/ViewModel/AccountScore/AccountScoreViewModel.swift:11]: check corectness

File: fearless/Modules/Banners/BannersViewModelFactory.swift [hunk 22]
```diff
@@ -53,11 +53,54 @@ final class BannersViewModelFactory: BannersViewModelFactoryProtocol {
                     subtitle: subtitle,
                     buttonTitle: buttonAction,
                     image: R.image.xorBanner()!,
-                    dismissable: true
+                    dismissable: true,
+                    fullsizeImage: false,
+                    bannerType: .buyXor
+                )
+            case .liquidityPools:
+                let title = "Liquidity pools"
```
Comment (bnsports) [inline fearless/Modules/Banners/BannersViewModelFactory.swift]: localization

File: fearless/Modules/Send/SendDependencyContainer.swift [hunk 23]
```diff
@@ -80,7 +80,7 @@ final class SendDepencyContainer {
             storageRequestPerformer: storageRequestPerformer
         )
 
-        cachedDependencies[chainAsset.uniqueKey(accountId: accountResponse.accountId)] = dependencies
+//        cachedDependencies[chainAsset.uniqueKey(accountId: accountResponse.accountId)] = dependencies
```
Comment (bnsports) [inline fearless/Modules/Send/SendDependencyContainer.swift]: commented code

File: fearless.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 24]
```diff
@@ -131,8 +140,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/soramitsu/shared-features-spm.git",
       "state" : {
-        "branch" : "fearless-wallet",
-        "revision" : "bb864277bcf22a6bd02d5abd7c8f07c5aa37236b"
+        "branch" : "network-module-update",
```
Comment (DRadmir) [inline fearless.xcworkspace/xcshareddata/swiftpm/Package.resolved]: should be merged in fearless wallet branch

File: fearless/ApplicationLayer/Services/AccountStatistics/Nomis/NomisAccountStatisticsFetcher.swift [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+        self.signer = signer
+    }
+}
+
+extension NomisAccountStatisticsFetcher: AccountStatisticsFetching {
+    func subscribeForStatistics(
+        address: String,
+        cacheOptions: CachedNetworkRequestTrigger
+    ) async throws -> AsyncThrowingStream<CachedNetworkResponse<AccountStatisticsResponse>, Error> {
+        guard let baseURL = URL(string: "https://api.nomis.cc/api/v1/multichain-score/wallet/") else {
```
Comment (DRadmir) [inline fearless/ApplicationLayer/Services/AccountStatistics/Nomis/NomisAccountStatisticsFetcher.swift]: Better use remote or ApplicationConfig

File: fearless/ApplicationLayer/Services/AccountStatistics/Nomis/NomisAccountStatisticsFetcher.swift [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+            address: address,
+            endpoint: "score"
+        )
+        request.signingType = .custom(signer: signer)
+        request.decoderType = .codable(jsonDecoder: NomisJSONDecoder())
+        return await networkWorker.performRequest(with: request, withCacheOptions: cacheOptions)
+    }
+
+    func fetchStatistics(address: String) async throws -> AccountStatisticsResponse? {
+        guard let baseURL = URL(string: "https://api.nomis.cc/api/v1/multichain-score/wallet/") else {
```
Comment (DRadmir) [inline fearless/ApplicationLayer/Services/AccountStatistics/Nomis/NomisAccountStatisticsFetcher.swift]: Better use remote or ApplicationConfig

File: fearless/ApplicationLayer/Services/ScamService/ScamInfoFetcher.swift [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+        let score: AccountStatisticsResponse? = try await accountScoreFetching.fetchStatistics(address: address)
+        let scamInfo: ScamInfo? = score.flatMap {
+            guard ($0.data?.score).or(.zero) < 0.25 else {
+                return nil
+            }
+
+            return ScamInfo(
+                name: "Nomis multi-chain score",
+                address: address,
+                type: .lowScore,
```
Comment (DRadmir) [inline fearless/ApplicationLayer/Services/ScamService/ScamInfoFetcher.swift:68]: why hard code?

File: fearless/ApplicationLayer/Services/ScamService/ScamInfoFetcher.swift [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+            return nil
+        }
+
+        return try? await fetchAccountScore(address: address)
+    }
+
+    // MARK: Private
+
+    private func fetchScamInfo(address: String) async throws -> ScamInfo? {
+        try await withCheckedThrowingContinuation { continuation in
```
Comment (DRadmir) [inline fearless/ApplicationLayer/Services/ScamService/ScamInfoFetcher.swift:42]: I propose use async repository

File: fearless/Common/Extension/ChainModel+Nomis.swift [hunk 29]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+import SSFModels
+
+extension ChainModel {
+    var isNomisSupported: Bool {
+        chainId == "137" || chainId == "1" || chainId == "56"
```
Comment (DRadmir) [inline fearless/Common/Extension/ChainModel+Nomis.swift:6]: Just append isNomisSupported in chain options

File: fearless/Common/Model/AccountStatistics.swift [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+    let totalRejectedTransactions: Int?
+    let averageTransactionTime: Decimal?
+    let maxTransactionTime: Decimal?
+    let minTransactionTime: Decimal?
+    let scoredAt: Date?
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        nativeBalanceUSD = try? container.decodeIfPresent(Decimal.self, forKey: .nativeBalanceUSD)
```
Comment (DRadmir) [inline fearless/Common/Model/AccountStatistics.swift:39]: the you use optional try?

File: fearless/Common/Model/ScamInfo.swift [hunk 31]
```diff
@@ -24,6 +24,7 @@ struct ScamInfo: Identifiable, Codable, Equatable, Hashable {
         case donation
         case exchange
         case sanctions
+        case lowScore = "Low network activity"
```
Comment (DRadmir) [inline fearless/Common/Model/ScamInfo.swift:27]: can we remove raw value?

File: fearless/Common/Services/ChainRegistry/ChainRegistry.swift [hunk 32]
```diff
@@ -226,14 +227,15 @@ final class ChainRegistry {
             return
         }
         chains.append(newChain)
-        _ = try ethereumConnectionPool.setupConnection(for: newChain)
+
+        _ = try? ethereumConnectionPool.setupConnection(for: newChain)
```
Comment (DRadmir) [inline fearless/Common/Services/ChainRegistry/ChainRegistry.swift]: use do catch and logger

File: fearless/Common/Services/ChainRegistry/ChainRegistry.swift [hunk 33]
```diff
@@ -226,14 +227,15 @@ final class ChainRegistry {
...
     }
 
     private func handleUpdatedEthereumChain(updatedChain: ChainModel) throws {
         guard let ethereumConnectionPool = self.ethereumConnectionPool else {
             return
         }
-        _ = try ethereumConnectionPool.setupConnection(for: updatedChain)
+        _ = try? ethereumConnectionPool.setupConnection(for: updatedChain)
```
Comment (DRadmir) [inline fearless/Common/Services/ChainRegistry/ChainRegistry.swift]: use do catch and logger

File: fearless/Common/ViewModel/AccountScore/AccountScoreViewModel.swift [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        Task {
+            do {
+                let stream = try await fetcher.subscribeForStatistics(address: address, cacheOptions: .onAll)
+                for try await statistics in stream {
+                    handle(response: statistics.value)
+                }
+            } catch {
+                print("Account statistics fetching error: ", error)
```
Comment (DRadmir) [inline fearless/Common/ViewModel/AccountScore/AccountScoreViewModel.swift]: replace print to logger

File: fearless/Modules/AccountStatistics/AccountStatisticsPresenter.swift [hunk 35]
```diff
@@ -0,0 +1,97 @@
...
+
+// MARK: - AccountStatisticsInteractorOutput
+
+extension AccountStatisticsPresenter: AccountStatisticsInteractorOutput {
+    func didReceiveAccountStatistics(_ response: AccountStatisticsResponse?) {
+        accountStatistics = response?.data
+        provideViewModel()
+    }
+
+    func didReceiveAccountStatisticsError(_: Error) {}
```
Comment (DRadmir) [inline fearless/Modules/AccountStatistics/AccountStatisticsPresenter.swift]: unused

File: fearlessTests/Modules/AccountStatistics/AccountStatisticsTests.swift [hunk 36]
```diff
@@ -0,0 +1,16 @@
+import XCTest
+
+class AccountStatisticsTests: XCTestCase {
```
Comment (DRadmir) [inline fearlessTests/Modules/AccountStatistics/AccountStatisticsTests.swift]: remove

**Review by DRadmir — CHANGES_REQUESTED** [review]

File: fearless/Common/Model/ScamInfo.swift (hunk 31 above)
Comment (bnsports) [reply fearless/Common/Model/ScamInfo.swift:27]: no

**Review by DRadmir — APPROVED** [review]

**Pull request merged**